File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        //base step
        //hypothesis step
        //induction step
        int res = INT_MIN;

        lengthOfTree(root, res);

return res - 1;
    }

    int lengthOfTree(TreeNode* root, int& res){

        if( root == NULL ){
            return 0;
        }

        int l = lengthOfTree(root->left, res);
        int r = lengthOfTree(root->right, res);

        int temp = max(l,r) + 1;
        int ans = max(temp, l+r+1);
         res = max(ans, res);

         return temp;
    }
        
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# bottom up dfs
def dfs(node, globalDia):

    # common area
    # leaf node
    if node.left is None and node.right is None:
        return 1

    # internal node
    leftDia, rightDia = 0, 0

    if node.left is not None:
        leftDia = dfs(node.left, globalDia)
        # localDiameter = leftDia + rightDia

    if node.right is not None:
        rightDia = dfs(node.right, globalDia)
        # localDiameter = leftDia + rightDia

    # before returning
    localDiameter = leftDia + rightDia
    globalDia[0] = max((localDiameter), globalDia[0]) 

    # globalDia[0] = max((leftDia + rightDia), globalDia[0]) 

    # return area
    if leftDia >= rightDia:
        # localDiameter = leftDia + rightDia
        return leftDia + 1
    else:
        # localDiameter = leftDia + rightDia
        return rightDia + 1
    


class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        # handle null tree
        if not root:
            return 0
        globalDia = [0]

        dfs(root, globalDia)
        return globalDia[0]
        
